dask/blockwise.py

```diff
@@ -591,21 +595,29 @@ def _dict(self):
 
             # Handle IO Subgraph
             for k in dsk:
-                io_key = (self.io_name,) + tuple([k[i] for i in range(1, len(k))])
-                if io_key in dsk[k]:
-                    # Inject IO-function arguments into the blockwise graph
-                    # as a single (packed) tuple.
-                    io_item = self.io_subgraph.get(io_key)
-                    io_item = list(io_item[1:]) if len(io_item) > 1 else []
-                    new_task = [io_item if v == io_key else v for v in dsk[k]]
-                    dsk[k] = tuple(new_task)
+                for io_name, io_subgraph in self.io_deps.items():
+                    # Leave out `new_axes` in key
+                    io_key = (io_name,) + tuple(
+                        [
+                            k[i + 1]
+                            for i, idx in enumerate(self.output_indices)
+                            if idx not in self.new_axes
+                        ]
+                    )
+                    if io_key in dsk[k]:
+                        # Inject IO-function arguments into the blockwise graph
+                        # as a single (packed) tuple.
+                        io_item = io_subgraph.get(io_key)
+                        io_item = list(io_item[1:]) if len(io_item) > 1 else []
+                        new_task = [io_item if v == io_key else v for v in dsk[k]]
+                        dsk[k] = tuple(new_task)
```

Comment:
This logic seems identical to a similar code block in `BlockwiseIO.__dask_distributed_unpack__`. There's a tradeoff here between the mental overhead of indirection and code duplication, but this seems complex enough to merit a separate method.

Reply:
Good point - There is no need to repeat this mess :)